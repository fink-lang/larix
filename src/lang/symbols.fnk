{symbol} = import '@fink/prattler/symbols'
{curr_loc} = import '@fink/prattler'

{next_is_new_expr} = import './block/indentation'
{single_expression} = import './block/expr'



infix_led = fn type, op, lbp: fn ctx, left:
  {loc: {start}} = left
  [right, next_ctx] = single_expression:: ctx, lbp
  {loc: {end}} = right

  [{type, op, left, right, loc: {start, end}}, next_ctx]


left_binding = fn lbp: fn ctx:
  match ctx:
    next_is_new_expr:: ?:
      0
    else:
      lbp


non_bining = fn: fn: 0


infix = fn op, type:
  {
    ...symbol:: op

    lbp: left_binding

    led: fn lbp: infix_led:: type, op, lbp + 1
  }


infix_right = fn op, type:
  {
    ...symbol:: op

    lbp: left_binding

    led: fn lbp: infix_led:: type, op, lbp - 1
  }


# TODO: why is this not type, op?
prefix = fn op, type:
  {
    ...symbol:: op

    lbp: non_bining

    nud: fn lbp: fn ctx:
      {start} = curr_loc:: ctx
      [right, next_ctx] = single_expression:: ctx, lbp
      {end} = right.loc

      [{type, op, right, loc: {start, end}}, next_ctx]
  }


literal = fn op, type:
  {
    ...symbol:: op

    lbp: non_bining

    nud: fn: fn ctx:
      loc = curr_loc:: ctx
      # value = curr_value:: ctx
      [{type, value: op, loc}, ctx]
  }


terminator = fn op:
  {
    ...symbol:: op

    lbp: left_binding
  }
