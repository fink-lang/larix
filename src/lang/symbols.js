import {symbol as parser_symbol} from '@fink/prattler/symbols';
import {next_loc} from '@fink/prattler';
import {indentation} from './indentation';


export const symbol = (id)=> ({
  ...parser_symbol(id),

  lbp: (lbp)=> (ctx)=> {
    if (next_loc(ctx).start.column <= indentation(ctx)) {
      return 0;
    }
    return lbp;
  }
});
