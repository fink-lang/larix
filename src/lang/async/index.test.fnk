{describe, it, expect, to_match_snapshot} = import '@fink/jest'

{parse_expr} = import '../../'


describe 'async', fn:

  it 'parses async: await foobar', fn:

    expect
      parse_expr 'await foobar'
      to_match_snapshot

