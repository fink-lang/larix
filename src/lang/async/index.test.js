import {other_token} from '@fink/prattler/symbols';

import {parse_expr} from '../../';


describe('assignment', ()=> {

  it('parses single line: foo = bar', ()=> {
    expect(
      parse_expr(`await foobar`)
    ).toMatchSnapshot();
  });
});
