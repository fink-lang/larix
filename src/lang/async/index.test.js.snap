// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assignment parses single line: foo = bar 1`] = `
await (1:0-1:12)
  ident (1:6-1:12) foobar
`;
